docs/madr/decisions/014-mesh-circuit-breaker-policy.md

```diff
@@ -0,0 +1,198 @@
+
+## Solution
+
+By dividing the configuration section into two separate, logical sections: `connectionPool` and `outlierDetection` it's
+easier to differentiate underlying envoy features without the need to split policy into two separate ones.
+
+Additional changes to the original `CircuitBreaker` policy consists of:
+
+- flattening `detectors.*` options with only single (`.consecutive`) property
+- changing word `Errors` to `Failures`
+- adding `origin` suffix to the local failures detector
+
+It means `totalErrors.consecutive`, `gatewayErrors.consecutive` and `localErrors.consecutive`
+becomes: `consecutiveTotalFailures`, `consecutiveGatewayFailures` and `consecutiveLocalOriginFailures` respectively.
+
+### Current configuration
+
+Below is the sample `MeshCircuitBreaker` configuration
+
+```yaml
+conf:
+  interval: 1s
+  baseEjectionTime: 30s
+  maxEjectionPercent: 20
+  splitExternalAndLocalErrors: false
+  thresholds:
+    maxConnections: 2
+    maxPendingRequests: 2
+    maxRequests: 2
+    maxRetries: 2
+  detectors:
+    totalErrors:
+      consecutive: 20
+    gatewayErrors:
+      consecutive: 10
+    localErrors:
+      consecutive: 7
+    standardDeviation:
+      requestVolume: 10
+      minimumHosts: 5
+      factor: 1.9
+    failure:
+      requestVolume: 10
+      minimumHosts: 5
+      threshold: 85
+
+```
+
+### Specification
+
+#### Top level
+
+As a part of this MADR root `targetRef`'s kind can be configured as one
+of `Mesh|MeshSubset|MeshService|MeshServiceSubset`. When we finally introduce `MeshHTTPRoute` kind and if there will be
+a reasonable use case, in the next iteration `MeshGatewayRoute` and `MeshHTTPRoute` could be added as a viable kinds.
+
+```yaml
+targetRef:
+  kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+  name: ...
+```
+
+#### From level
+
+`MeshCircuitBreaker` is outbound only policy, so only `to` should be configured.
+
+#### To level
+
+```yaml
+to:
+- targetRef:
+    kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+    name: ...
+```
+
+#### Minimal configuration
+
+As all the connection pool related properties have default values, there is no need to specify any of these to apply the
+policy
+
+```yaml
+default:
+  connectionPool: { }
+```
+
+### Examples
+
+#### Default configuration
+
+##### `MeshCircuitBreaker`
+
+```yaml
+type: MeshCircuitBreaker
+mesh: default
+name: default-circuit-breaker
+spec:
+  targetRef:
+    kind: Mesh
+    name: default
+  to:
+  - targetRef:
+      kind: Mesh
+      name: default
+    default:
+      connectionPool:
+        maxConnections: 1024
+        maxPendingRequests: 1024
+        maxRetries: 3
+        maxRequests: 1024
+```
+
+#### Extensive configuration
+
+```yaml
+type: MeshCircuitBreaker
+mesh: default
+name: extensive-circuit-breaker
+spec:
+  targetRef:
+    kind: Mesh
+    name: default
+  to:
+  - targetRef:
+      kind: Mesh
+      name: default
+    default:
+      connectionPool:
```

Comment:
In practice there can be multiple thresholds in Envoy I'm not sure what the point of this would be but have we thought about it?

Reply:
It's indeed true, but we are not setting priorities on routes, and I'm on a fence to be honest, as introducing thresholds with only one accepted priority i.e.:
```yaml
thresholds:
- priority: DEFAULT
  maxConnections: 1024
  maxPendingRequests: 1024
  maxRetries: 3
  maxRequests: 1024
```
would definitely be more future proof, but if it's not likely that we will be using it makes the UX much worse as operators might be confused what does the `priority` currently mean, why it's a collection which accepts only one item etc.

Reply:
I've spend some time on this issue and adjusted the MADR to make this more feature-proof, without compromising the UX very much I believe